lib/mongo/client.rb

```diff
@@ -855,14 +857,32 @@ def validate_new_options!(opts = Options::Redacted.new)
     end
 
     # Validates all options after they are set on the client.
-    # This method is intended to catch combinations of options which do are
+    # This method is intended to catch combinations of options which are
     # not allowed.
     def validate_options!
       if options[:write] && options[:write_concern] && options[:write] != options[:write_concern]
         raise ArgumentError, "If :write and :write_concern are both given, they must be identical: #{options.inspect}"
       end
     end
 
+    # Validates all authentication-related options after they are set on the client
+    # This method is intended to catch combinations of options which are not allowed
+    def validate_authentication_options!
+      case options[:auth_mech]
+      when :mongodb_cr, :plain, :scram, :scram256
+        raise Mongo::Auth::InvalidConfiguration.new(options[:auth_mech], 'user is required') if options[:user].nil?
+      when :mongodb_x509
+        raise Mongo::Auth::InvalidConfiguration.new(:mongodb_x509, 'auth_source must be $external or nil') if !['$external', nil].include?(options[:auth_source])
+        raise Mongo::Auth::InvalidConfiguration.new(:mongodb_x509, 'password is not supported') if !options[:password].nil?
+      when nil
+        raise Mongo::Auth::InvalidConfiguration.new(:default, 'auth_source cannot be present without a user') if !options[:auth_source].nil? && options[:user].nil?
+        raise Mongo::Auth::InvalidConfiguration.new(:default, 'do not include user information delimiter (@) in URI for blank user') if !options[:user].nil? && !options[:user].length
+        raise Mongo::Auth::InvalidConfiguration.new(:default, 'do not include password delimiter (:) in URI for blank password') if !options[:password].nil? && !options[:password].length
```

Comment:
These last two cases are related to parsing a connection string with user information but without a username or password (e.g. `"mongodb://@localhost.com/"`).

In the cases where no auth mechanism is specified, I'm passing `:default` into the error class. Let me know if you think this is clear enough.